### ISSUE TITLE
WP Rocket’s container can trigger an error

### ISSUE THREAD

**Issue by Screenfeed** [wp-media/imagify-plugin#485]
When we try to get a WP Rocket’s container’s unused entry, the container throws an error. It should be catched.

```
An error of type E_ERROR was caused in line 266 of the file /public_html/wp-content/plugins/wp-rocket/vendor/league/container/src/Container.php. Error message: Uncaught League\Container\Exception\NotFoundException: Alias (cdn) is not being managed by the container in /public_html/wp-content/plugins/wp-rocket/vendor/league/container/src/Container.php:266
Stack trace:
#0 /public_html/wp-content/plugins/wp-rocket/vendor/league/container/src/Container.php(93): League\Container\Container->getFromDelegate('cdn', Array)
#1 /public_html/wp-content/plugins/imagify/inc/3rd-party/wp-rocket/classes/Main.php(82): League\Container\Container->get('cdn')
#2 /public_html/wp-includes/class-wp-hook.php(288): Imagify\ThirdParty\WPRocket\Main->set_cdn_source(Array)
#3 /public_html/wp-includes/plugin.php(206): WP_Hook->apply_filters(Array, Array)
#4 /public_html/wp-content/plugins/imagify/classes/Webp/Picture/Display.php(689): apply_filters('imagify_cdn_sou...', Array)
#5 /public_html/wp-content/plugins/imagify/views/part-settings-webp.php(45): Imagify\Webp\Picture\Display->get_cdn_source()
```

**Comment by Miraeld** [comment]
> [!NOTE]
> Generated by the AI delivery pipeline (grooming-agent · claude-opus-5). **Edited 2026-07-27 to correct a factual error — see "Correction" below.**

### Grooming Plan — Issue #485 (Plan v2)

**Approach:** Harden the WP Rocket container resolution inside `Imagify\ThirdParty\WPRocket\Main::set_cdn_source()` so it falls through to the existing `get_rocket_cdn_cnames()` fallback instead of fataling.
**Effort:** XS · **Risk:** LOW · **Complexity:** LOW

---

### ⚠️ Correction to Plan v1

**Plan v1 of this comment stated: _"WP Rocket does not register it in every context, so `get( 'cdn' )` throws `NotFoundException`."_ That claim is wrong, and was asserted without verification against WP Rocket's source.**

`cdn` is registered **unconditionally** in every WP Rocket version checked, including the version current when this issue was filed in March 2020:

| WP Rocket version | Registration site | Conditional? |
|---|---|---|
| v3.3 | `inc/classes/class-plugin.php:110` → `Common_Subscribers` | No |
| v3.4.4 | `inc/classes/class-plugin.php:119` | No |
| **v3.5.1** (current when this issue was filed) | `class-plugin.php:124`, `class-common-subscribers.php:66` | No |
| current `develop` | `inc/Plugin.php:287` → `inc/Engine/CDN/ServiceProvider.php:63` | No |

In each case the provider is added at method-body top level — not gated on the `cdn` option, not on `is_admin()`, not on license validity. (Compare `CloudflareServiceProvider`, which *is* gated, on `do_cloudflare`. CDN is not.)

**Consequence: the mechanism this issue describes is not reproducible, and never was.** The original report is backed by a genuine stack trace, so something real happened — but the option-gate explanation is ruled out by source inspection.

The one genuine structural window: `rocket_container` is hooked in `Plugin::__construct()` (`inc/Plugin.php:116`), but `cdn` is not registered until `load()` (`inc/Plugin.php:287`). Between those, the filter returns a live-but-empty container and `method_exists( $container, 'get' )` passes trivially. **Imagify cannot reach that window** — all three appliers of `imagify_cdn_source_url` run long after `plugins_loaded`. So this does not explain the 2020 crash either.

**The 2020 crash mechanism remains unexplained.** Rather than invent one, the work below is scoped as hardening of an unsafe assumption that is demonstrably present in Imagify's code, not as a fix for a reproducible bug.

---

### Why still worth doing

The residual risk is version-independent and does not depend on reproducing this issue:

1. `rocket_container` is a **public filter** — any third-party plugin can return a partial, foreign, or non-container object. Imagify cannot control this and WP Rocket cannot fix it upstream.
2. `get()` can throw even when `has()` is true: league/container raises `ContainerException` if a provider claims a service in `provides()` but fails to define it on `register()`, or if the definition's own dependencies (`options`) fail to resolve.
3. `CDN::get_cdn_urls()` is third-party code in the same block and can throw independently, via `cname_validator->is_valid()`.

**Blast radius.** `imagify_cdn_source_url` has three appliers, all reachable today:

| Call site | Trigger |
|---|---|
| `views/part-settings-webp.php:89` | Settings page render — unconditional, not gated on the next-gen checkbox |
| `inc/classes/class-imagify-options.php:183` | Saving Imagify options |
| `classes/Picture/Display.php:735` | **Frontend page load**, during `<picture>` rewriting |

The third was not in the original trace and is the strongest argument for guarding: an uncaught throw there fatals **public pages**, not just the admin screen. The `severity: minor` label is understated on impact, though defensible on frequency.

### Mechanism: `has()` vs `try/catch`

Verified against the container WP Rocket actually ships: **league/container 4.2.5**, Strauss-prefixed to `WP_Rocket\Dependencies\League\Container\Container`. Its `has()` checks definitions, tags, service providers **and** delegates, and `resolve()` checks the same four sources in the same order before throwing — so `has( 'cdn' ) === true` reliably implies `get( 'cdn' )` will not throw `NotFoundException`.

**Recommendation: `try { } catch ( \Throwable $e )` with `has()` as a cheap pre-check inside it.** `has()` alone closes the reported fatal but not the dependency-failure class (point 2 above). `\Throwable` rather than `\Exception` because a foreign container can raise an `Error`.

### Target shape

```php
$container = apply_filters( 'rocket_container', null );

if ( is_object( $container ) && method_exists( $container, 'get' ) ) {
	try {
		if ( ! method_exists( $container, 'has' ) || $container->has( 'cdn' ) ) {
			$cdn = $container->get( 'cdn' );

			if ( $cdn && method_exists( $cdn, 'get_cdn_urls' ) ) {
				$url = $cdn->get_cdn_urls( [ 'all', 'images' ] );
			}
		}
	} catch ( \Throwable $e ) {
		$url = null;
	}
}
```

`$url = null` (not an empty catch) keeps the existing `if ( ! isset( $url ) && function_exists( 'get_rocket_cdn_cnames' ) )` branch working untouched — `isset( null )` is `false`. Everything after that block stays byte-identical, so the happy path is provably unchanged.

### Layer decision

Fix in place in the legacy file. `Main` lives in `inc/3rd-party/` and uses `InstanceGetterTrait`; migrating it to `classes/ThirdParty/Plugins/WPRocket.php` as a `SubscriberInterface` is the cleaner end state, but it changes *when* the filter is registered and should not ride along with a defensive fix. Separate follow-up.

### Tests

Unit test `Tests/Unit/inc/3rd-party/wp-rocket/Main/setCdnSource.php`, Brain Monkey, container doubles as anonymous classes so the suite never depends on WP Rocket being installed.

1. `has( 'cdn' )` false, `get()` throws if reached → falls back, nothing escapes.
2. `has()` true but `get()` throws → contained, falls back.
3. `get_cdn_urls()` throws → contained, falls back.
4. Happy path: container resolves → URL comes from the container, not the CNAME helper.
5. Container throws **and** `get_rocket_cdn_cnames()` empty → original `$source` returned untouched.

**Test command:** `composer test-unit -- --filter="Test_SetCdnSource"`
Plus `composer phpcs -- inc/3rd-party/wp-rocket/classes/Main.php && composer run-stan`.

### Out of scope

The migration above; `is_array( $url )` hardening before `reset()`; logging the caught exception; any change to `classes/CDN/CDN.php`, `classes/Picture/Display.php` or the view. No frontend work.

### Recommended disposition

Merge #1200 as hardening. **Close this issue as not reproducible**, referencing the PR — not as "fixed", which would overstate what was established.
